### PR TITLE
Remove MessageContainer from ParseContext + storage type change

### PR DESCRIPTION
**Pull request by jokva**
As a step towards (re)making `ParseContext` a read-only configuration object, remove `MessageContainer` from parse context. The patch set also changes Messages to auto-only storage, deprecating `unique_ptr`s.

### CODE REVIEW

**Comment by akva2** [comment]
Can one of the admins verify this patch?

File: opm/parser/eclipse/Parser/MessageContainer.hpp [hunk 1]
```diff
@@ -43,21 +41,31 @@ namespace Opm {
 
 
     struct Location {
+        Location() = default;
+        Location( const std::string& s, int l ) : filename( s ), lineno( l ) {}
```
Comment (atgeirr) [inline opm/parser/eclipse/Parser/MessageContainer.hpp]: `l` is a bad variable name, being too close to a `1`.

File: opm/parser/eclipse/Parser/MessageContainer.cpp [hunk 2]
```diff
@@ -25,94 +25,127 @@
...
 
 
-    void MessageContainer::error(const std::string& msg, const std::string& filename, const int lineno)
-    {
-        std::unique_ptr<Location> loc(new Location{filename, lineno});
-        m_messages.emplace_back(Message{MessageType::MessageTypeEnum::Error, msg, std::move(loc)});
+    void MessageContainer::error( const std::string& msg,
+                                  const std::string& filename,
+                                  const int lineno ) {
+        m_messages.emplace_back( Message {
```
Comment (atgeirr) [inline opm/parser/eclipse/Parser/MessageContainer.cpp]: Is not one of the reasons to use `emplace_back` that you can drop the explicit construction and just pass the arguments?
```
m_messages.emplace_back(
    MessageType::MessageTypeEnum::Error,
    msg,
    { filename, lineno });
```

File: opm/parser/eclipse/Parser/tests/MessageContainerTest.cpp [hunk 3]
```diff
@@ -34,8 +34,8 @@ BOOST_AUTO_TEST_CASE(TestIterator) {
     msgContainer.bug("This is a bug.", "dummy.log", 20);
     {       
         BOOST_CHECK_EQUAL("This is an error.", msgContainer.begin()->message);
-        BOOST_CHECK_EQUAL("dummy.log", (msgContainer.end()-1)->location->filename);
-        BOOST_CHECK_EQUAL(20, (msgContainer.end()-1)->location->lineno);
+        BOOST_CHECK_EQUAL("dummy.log", (msgContainer.end()-1)->location.filename);
+        BOOST_CHECK_EQUAL(20, (msgContainer.end()-1)->location.lineno);
     }
```
Comment (joakim-hove) [inline opm/parser/eclipse/Parser/tests/MessageContainerTest.cpp]: Can we have a test demonstrating the `operator bool( )` functionality of the `Location` class.

File: opm/parser/eclipse/Parser/MessageContainer.cpp (hunk 2 above)
Comment (jokva) [inline opm/parser/eclipse/Parser/MessageContainer.cpp]: The compiler is unable to deduce it correctly without the hint from `Message {}` (or constructor definitions). I use `emplace_back` just to emphasise the move going on.

An alternative is surely to define two constructors for `Message`, in which case the `Message {}` can go away.

File: opm/parser/eclipse/Parser/tests/MessageContainerTest.cpp (hunk 3 above)
Comment (jokva) [inline opm/parser/eclipse/Parser/tests/MessageContainerTest.cpp]: Yes.

File: opm/parser/eclipse/Parser/MessageContainer.hpp (hunk 1 above)
Comment (jokva) [inline opm/parser/eclipse/Parser/MessageContainer.hpp]: Actually a slight oversight on my part, this was mostly thrown in for testing the emplace_back syntax in https://github.com/OPM/opm-parser/pull/748#discussion_r58354804

I

**Comment by joakim-hove** [comment]
jenkins build this please

File: opm/parser/eclipse/Parser/MessageContainer.hpp [hunk 4]
```diff
@@ -20,17 +20,27 @@
...
 #include <vector>
 #include <memory>
 
 namespace Opm {
 
-    namespace MessageType {
+    struct Location {
+        Location() = default;
+        Location( const std::string& fn, int ln ) :
+            filename( fn ), lineno( ln ) {}
```
Comment (joakim-hove) [inline opm/parser/eclipse/Parser/MessageContainer.hpp]: This should throw for `ln <  0` - it should be freaking impossible to explicitly create a `Location` instance which evaluates to `false`.

**Comment by joakim-hove** [comment]
I will merge this when we have a test demonstrating the `bool operator()`behavior of the `class Location` - something like:
```C++
Location loc_true("filename" , 100);
Location loc_false;
BOOST_CHECK( loc_true );
BOOST_CHECK( !loc_false);
BOOST_CHECK_THROW( Location("filename" , -1) , std::invalid_argument);
```

**Comment by jokva** [comment]
@joakim-hove pushed, but without throw check. Simply because location.filename should be considered undefined if lineno is nonsensical.

Obviously this makes  container.warning( msg, filename, 0 ) which I'm happy to disallow.

**Comment by joakim-hove** [comment]
Well - my point was to disallow construction of  an invalid location by checking and throwing in the cobstructor,  and documenting that with a test.

File: opm/parser/eclipse/Parser/MessageContainer.hpp (hunk 4 above)
Comment (blattms) [inline opm/parser/eclipse/Parser/MessageContainer.hpp]: On Tue, Apr 05, 2016 at 01:45:05PM -0700, Joakim Hove wrote:
> >  #include <string>
> >  #include <vector>
> >  #include <memory>
> >  
> >  namespace Opm {
> >  
> > -    namespace MessageType {
> > +    struct Location {
> > +        Location() = default;
> > +        Location( const std::string& fn, int ln ) :
> > +            filename( fn ), lineno( ln ) {}
> 
> This should throw for `ln <  0` - it should be freaking impossible to explicitly create a `Location` instance which evaluates to `false`.

or instead change the type from int to std::size_t or another
unsgigned integer as negative values make no sense at all

**Comment by joakim-hove** [comment]
OK - thank you for the test update.

**Pull request merged**